File: clients/apple/Packages/MailcalKit/Sources/MailcalUI/AvatarView.swift
import MailcalBindings
import SwiftUI

#if canImport(AppKit)
    import AppKit
    private typealias PlatformImage = NSImage
#else
    import UIKit
    private typealias PlatformImage = UIImage
#endif

/// The circle drawn beside a person: their photo when one is known, else their monogram on the
/// colour the core picked for them.
///
/// **Only the shape is decided here.** The letters, the colour and whether there is a photo all
/// come from the core (`docs/avatars.md`), resolved per client, four clients would disagree
/// about whether a white letter is legible on a mid-green fill, exactly as they would about a
/// calendar chip's label.
struct AvatarView: View {
    let avatar: Avatar
    var diameter: CGFloat = 34

    @Environment(\.colorScheme) private var colorScheme
    @State private var photo: PlatformImage?

    private var swatch: Swatch { colorScheme == .dark ? avatar.dark : avatar.light }

    var body: some View {
        Circle()
            .fill(parseHexColor(swatch.background))
            .frame(width: diameter, height: diameter)
            .overlay { content }
            // The photo is drawn to the circle's edge, so it is clipped rather than inset.
            .clipShape(Circle())
            // Decoration. The row already announces the person's name, and a monogram is a
            // restatement of it, announcing this would make VoiceOver read a letter before
            // every sender. The person glyph says nothing at all.
            .accessibilityHidden(true)
            .task(id: avatar.imagePath) { await loadPhoto() }
    }

    @ViewBuilder
    private var content: some View {
        if let photo {
            Image(platformImage: photo)
                .resizable()
                .scaledToFill()
                .frame(width: diameter, height: diameter)
        } else if avatar.initials.isEmpty {
            // Neither a name nor an address. The core deliberately sends no placeholder text:
            // any word it chose would be untranslatable English, so the platform's own glyph
            // stands in.
            Image(systemName: "person.crop.circle")
                .foregroundStyle(parseHexColor(swatch.text))
        } else {
            Text(avatar.initials)
                .font(.system(size: diameter * 0.4, weight: .medium))
                .foregroundStyle(parseHexColor(swatch.text))
        }
    }

    private func loadPhoto() async {
        guard let path = avatar.imagePath else {
            photo = nil
            return
        }
        photo = await AvatarPhotoCache.shared.image(atPath: path)
    }
}

/// Decoded avatar photos, keyed by their file path.
///
/// The path is safe as a cache key precisely because the engine names the file by a hash of its
/// own contents: a changed photo is a changed path, so an entry can never go stale and nothing
/// needs invalidating.
private actor AvatarPhotoCache {
    static let shared = AvatarPhotoCache()

    private let cache = NSCache<NSString, PlatformImage>()

    /// The decoded image at `path`, or `nil` if it cannot be read.
    ///
    /// Decoding happens off the main actor: a list scrolls past dozens of these, and doing it
    /// inline would drop frames. The core has already checked the bytes are a raster image
    /// within a size cap, so this does not sniff again.
    func image(atPath path: String) -> PlatformImage? {
        let key = path as NSString
        if let cached = cache.object(forKey: key) { return cached }
        guard let image = PlatformImage(contentsOfFile: path) else { return nil }
        cache.setObject(image, forKey: key)
        return image
    }
}

extension Image {
    fileprivate init(platformImage: PlatformImage) {
        #if canImport(AppKit)
            self.init(nsImage: platformImage)
        #else
            self.init(uiImage: platformImage)
        #endif
    }
}

/// The unread marker on a desktop-layout row, and the space it reserves when read.
///
/// A **dot**, not the envelope glyph it replaces: the avatar now occupies the gutter, and two
/// symbols competing for the same job left the row saying "unread" twice. `docs/avatars.md` binds the
/// dot to desktop layouts only, the compact phone list has no room for it, and bold subject
/// and sender already carry unread there.
/// **Whether to draw one is not decided here**, see `ContentView.unreadDot(_:)`, which is the
/// only thing that should build this. The size class has to be read at the window level, and a
/// row is the wrong place to read it from.
struct UnreadDot: View {
    let unread: Bool

    var body: some View {
        Circle()
            .fill(unread ? Color.accentColor : Color.clear)
            .frame(width: 7, height: 7)
            // The row's own accessibility value carries read state; a second announcement of it
            // before every sender is noise.
            .accessibilityHidden(true)
    }
}

File: clients/windows/Mailcal/Controls/AvatarView.cs
// The circle a row draws beside a person: their photo when a synced address book has one, else
// their initials on the colour the core picked for them (docs/avatars.md).
//
// **Only the shape is decided here.** The letters and the colour come from the core, for the same
// reason a calendar chip's contrast does: resolved per client, four clients disagree about whether
// a white letter is legible on a mid-green fill. So this reuses CalendarColors.Parse rather than
// growing a second hex reader, the two surfaces draw from one palette on purpose.
//
// A Grid rather than a templated Control, like ColumnSplitter: an Ellipse with one centred child
// needs no control template, and filling the Ellipse with an ImageBrush clips a photo to the circle
// by construction instead of depending on how a Border treats its corner radius.

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Allodia.Mailcal.Calendar;
using Allodia.Mailcal.Services;
using Allodia.Mailcal.ViewModels;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Automation.Peers;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Media.Imaging;
using Microsoft.UI.Xaml.Shapes;

namespace Allodia.Mailcal.Controls;

/// <summary>The avatar circle: a person's photo, else their monogram on their own colour.</summary>
public sealed class AvatarView : Grid
{
    // Segoe Fluent Icons "Contact", the platform's own person glyph, for a row that names nobody.
    private const string PersonGlyph = "";

    // Photos decode once at this height and every size draws from that one bitmap: the largest
    // avatar on screen is 56 effective pixels, which is 112 physical at 200% scale.
    private const int PhotoDecodePixels = 128;

    // A photo's file name is a hash of its own contents, so an entry can never go stale and there
    // is nothing to invalidate. Cleared wholesale at the cap rather than aged out, an LRU would be
    // more machinery than a few hundred small bitmaps deserve. UI-thread only, like every control.
    private const int PhotoCacheCap = 256;

    private static readonly Dictionary<string, BitmapImage> Photos = new();

    // Paths that failed to decode, so a recycled row does not retry the same broken file, and the
    // log records it once rather than once per scroll.
    private static readonly HashSet<string> Unreadable = new();

    // The theme the circle was last drawn in, see the Loaded handler below.
    private bool _drawnDark;

    /// <summary>Identifies the <see cref="Avatar"/> property.</summary>
    public static readonly DependencyProperty AvatarProperty = DependencyProperty.Register(
        nameof(Avatar), typeof(AvatarItem), typeof(AvatarView), new PropertyMetadata(null, OnChanged));

    /// <summary>Identifies the <see cref="Diameter"/> property.</summary>
    public static readonly DependencyProperty DiameterProperty = DependencyProperty.Register(
        nameof(Diameter), typeof(double), typeof(AvatarView), new PropertyMetadata(34.0, OnChanged));

    /// <summary>Creates the circle. It draws nothing until an <see cref="Avatar"/> is set.</summary>
    public AvatarView()
    {
        // Decoration, on every platform. The row already announces the person's name and the
        // monogram restates its first letters, so announcing it would make Narrator read a letter
        // before every sender; the person glyph says nothing at all.
        AutomationProperties.SetAccessibilityView(this, AccessibilityView.Raw);
        // The core resolved both themes up front, so a light/dark switch is a repaint here and
        // never a trip back through the FFI.
        ActualThemeChanged += (_, _) => Apply();
        // A row's template fills its bindings before the container joins the tree, where
        // ActualTheme is still the default rather than the window's. Redrawing on Loaded catches
        // that; the guard keeps it to the rows that were actually drawn in the wrong theme, so it
        // is not a second rebuild per row.
        Loaded += (_, _) =>
        {
            if (_drawnDark != (ActualTheme == ElementTheme.Dark))
            {
                Apply();
            }
        };
    }

    /// <summary>The person this circle is of.</summary>
    public AvatarItem? Avatar
    {
        get => (AvatarItem?)GetValue(AvatarProperty);
        set => SetValue(AvatarProperty, value);
    }

    /// <summary>The circle's diameter in effective pixels.</summary>
    public double Diameter
    {
        get => (double)GetValue(DiameterProperty);
        set => SetValue(DiameterProperty, value);
    }

    private static void OnChanged(DependencyObject element, DependencyPropertyChangedEventArgs args) =>
        ((AvatarView)element).Apply();

    private void Apply()
    {
        Children.Clear();
        Width = Diameter;
        Height = Diameter;
        if (Avatar is not { } avatar)
        {
            return;
        }

        var dark = ActualTheme == ElementTheme.Dark;
        _drawnDark = dark;
        var circle = new Ellipse
        {
            Width = Diameter,
            Height = Diameter,
            Fill = new SolidColorBrush(CalendarColors.Parse(avatar.Background(dark))),
        };
        Add(circle);

        var ink = new SolidColorBrush(CalendarColors.Parse(avatar.TextColor(dark)));
        // A photo is drawn OVER the monogram rather than instead of it: decoding is asynchronous,
        // and the contract says an avatar is never blank, so the letters hold the circle until the
        // face arrives, and stay if it never does.
        var monogram = avatar.Content == AvatarContent.PersonGlyph
            ? Glyph(ink)
            : Letters(avatar.Initials, ink);
        Add(monogram);

        if (avatar.Content == AvatarContent.Photo)
        {
            _ = ShowPhotoAsync(circle, monogram, avatar.ImagePath!);
        }
    }

    private FrameworkElement Letters(string initials, Brush ink) => new TextBlock
    {
        Text = initials,
        FontSize = Diameter * 0.4,
        FontWeight = Microsoft.UI.Text.FontWeights.Medium,
        Foreground = ink,
        HorizontalAlignment = HorizontalAlignment.Center,
        VerticalAlignment = VerticalAlignment.Center,
    };

    private FrameworkElement Glyph(Brush ink) => new FontIcon
    {
        Glyph = PersonGlyph,
        FontSize = Diameter * 0.5,
        Foreground = ink,
        HorizontalAlignment = HorizontalAlignment.Center,
        VerticalAlignment = VerticalAlignment.Center,
    };

    private void Add(FrameworkElement child)
    {
        // Raw on the child too: the flag does not inherit, and it is the TextBlock, not the panel
        // around it, that a screen reader would otherwise land on and read out.
        AutomationProperties.SetAccessibilityView(child, AccessibilityView.Raw);
        Children.Add(child);
    }

    // Swap the fill for the photo once it has decoded. Nothing waits on a face, so a row that
    // scrolled away before it landed simply drops the result.
    private async Task ShowPhotoAsync(Ellipse circle, FrameworkElement monogram, string path)
    {
        var photo = await PhotoAsync(path);
        if (photo is null)
        {
            return; // The monogram stands. Never a blank circle.
        }
        // The container may have been recycled onto another row while this decoded, which would
        // otherwise put one person's face on someone else's mail.
        if (Avatar?.ImagePath != path)
        {
            return;
        }
        circle.Fill = new ImageBrush { ImageSource = photo, Stretch = Stretch.UniformToFill };
        monogram.Visibility = Visibility.Collapsed;
    }

    private static async Task<BitmapImage?> PhotoAsync(string path)
    {
        if (Photos.TryGetValue(path, out var cached))
        {
            return cached;
        }
        if (Unreadable.Contains(path))
        {
            return null;
        }
        try
        {
            var photo = new BitmapImage { DecodePixelHeight = PhotoDecodePixels };
            using var file = File.OpenRead(path);
            await photo.SetSourceAsync(file.AsRandomAccessStream());
            if (photo.PixelHeight == 0)
            {
                // A decode that fails raises ImageFailed rather than throwing, so the awaited call
                // can return having produced nothing at all.
                throw new InvalidOperationException("the file decoded to no pixels");
            }
            if (Photos.Count >= PhotoCacheCap)
            {
                Photos.Clear();
            }
            Photos[path] = photo;
            return photo;
        }
        catch (Exception error)
        {
            Unreadable.Add(path);
            Log.Warn($"a contact photo could not be drawn: {error.GetType().Name}");
            return null;
        }
    }
}
